pkg/agent/flexvolume/manager/ceph/manager.go

```diff
@@ -0,0 +1,180 @@
+/*
+Copyright 2016 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+Some of the code below came from https://github.com/coreos/etcd-operator
+which also has the apache 2.0 license.
+*/
+package ceph
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+	"time"
+
+	cephmon "github.com/rook/rook/pkg/ceph/mon"
+	"github.com/rook/rook/pkg/model"
+	"github.com/rook/rook/pkg/util/sys"
+
+	"github.com/coreos/pkg/capnslog"
+	cephclient "github.com/rook/rook/pkg/ceph/client"
+	"github.com/rook/rook/pkg/clusterd"
+	"github.com/rook/rook/pkg/operator/mon"
+)
+
+var logger = capnslog.NewPackageLogger("github.com/rook/rook", "ceph-volumeattacher")
+
+// VolumeManager represents an object for perform volume attachment requests for Ceph volumes
+type VolumeManager struct {
+	context          *clusterd.Context
+	devicePathFinder model.DevicePathFinder
+}
+
+type devicePathFinder struct{}
+
+// NewVolumeManager create attacher for ceph volumes
+func NewVolumeManager(context *clusterd.Context) *VolumeManager {
+	return &VolumeManager{
+		context:          context,
+		devicePathFinder: &devicePathFinder{},
+	}
+}
+
+// Attach a ceph image to the node
+func (vm *VolumeManager) Attach(image, pool, clusterName string) (string, error) {
+
+	// check if the volume is already attached
+	devicePath, err := vm.isAttached(image, pool, clusterName)
+	if err != nil {
+		return "", fmt.Errorf("failed to check if volume %s/%s is already attached: %+v", pool, image, err)
+	}
+	if devicePath != "" {
+		logger.Infof("volume %s/%s is already attached. The device path is %s", pool, image, devicePath)
+		return devicePath, nil
+	}
+
+	// attach and poll until volume is mapped
+	logger.Infof("attaching volume %s/%s cluster %s", pool, image, clusterName)
+	monitors, keyring, err := getClusterInfo(vm.context, clusterName)
+	defer os.Remove(keyring)
+	if err != nil {
+		return "", fmt.Errorf("failed to load cluster information from cluster %s: %+v", clusterName, err)
+	}
+
+	err = cephclient.MapImage(vm.context, image, pool, clusterName, keyring, monitors)
+	if err != nil {
+		return "", fmt.Errorf("failed to map image %s/%s cluster %s. %+v", pool, image, clusterName, err)
+	}
+
+	// poll for device path
+	retryCount := 0
+	maxRetries := 10
```

Comment:
is 10s of retrying sufficient? And would be good to define in a const

Reply:
it is not 10s. It is 10 retries. I got this code from the rookctl map. Rookctl also tries 10 times. I can make it a const.
https://github.com/rook/rook/blob/master/cmd/rookctl/block/map.go#L128